### PR TITLE
fix: don't treat \ as escape in standard strings, support E-strings, support vars after ->> operator, treat lowercase e as valid e-string prefix

### PR DESCRIPTION
**Pull request by sdepold**
Backport of #14700 and #14733.

### CODE REVIEW

**Review by WikiRik — COMMENTED** [review]
It feels a bit stupid, but can you undo the various style changes? Makes it easier to review the changes, especially in the tests.

Apart from that it looks good to me

**Comment by sdepold** [comment]
> It feels a bit stupid, but can you undo the various style changes? Makes it easier to review the changes, especially in the tests.
> 
> Apart from that it looks good to me

to be completely honest with you, I think we should stop wasting time talking about formatting :D we have the linter in place to prevent syntax problems. it even auto-formats stuff as it see needs. it's all green. so why care about something that is legacy anyway. however, I've reverted the changes to best of my abilities, even preventing the auto-formatting of my editor to turn it all upside down again.

File: test/unit/utils/sql.test.js [hunk 1]
```diff
@@ -44,6 +62,18 @@ describe('injectReplacements (named replacements)', () => {
...
+  it('parses bind parameters following JSONB indexing', () => {
+    const sql = injectReplacements(
+      'SELECT * FROM users WHERE json_col->>$key',
+      dialect,
+      { id: 1 }
+    );
+
+    expectsql(sql, {
+      default: 'SELECT * FROM users WHERE json_col->>$key'
+    });
```
Comment (WikiRik) [inline test/unit/utils/sql.test.js:74]: Why are these commented out?

File: test/unit/utils/sql.test.js [hunk 2]
```diff
@@ -109,14 +139,83 @@ describe('injectReplacements (named replacements)', () => {
...
+        default:
+          new Error(`The following SQL query includes an unterminated string literal:
+SELECT * FROM users WHERE id = E'\\' $id' OR id = $id`),
+        'mysql mariadb': toMatchSql(
+          "SELECT * FROM users WHERE id = E'\\' $id' OR id = $id"
+        ),
+        postgres: "SELECT * FROM users WHERE id = E'\\' $id' OR id = $id"
+      }
+    );
+  });
```
Comment (WikiRik) [inline test/unit/utils/sql.test.js:199]: What is the difference between mysql/mariadb and postgres?

Comment (sdepold) [reply test/unit/utils/sql.test.js:199]: probably none :) there was a diff in the original version but that was due to a difference in functionality where the function would return bind values. let me adjust this case

**Review by WikiRik — APPROVED** [review]

**Review by ephys — APPROVED** [review]

**Pull request merged**